core/services/log/broadcaster.go

```diff
@@ -25,92 +31,528 @@ type (
 		utils.DependentAwaiter
 		Start() error
 		Stop() error
-		Register(contract AbigenContract, listener Listener) (connected bool)
-		Unregister(contract AbigenContract, listener Listener)
+		Register(listener Listener, opts ListenerOpts) (connected bool, unsubscribe func())
 	}
 
 	broadcaster struct {
-		subscriber *subscriber
-		relayer    *relayer
+		orm       ORM
+		ethClient eth.Client
+		config    Config
+		connected *abool.AtomicBool
+
+		registrations map[common.Address]map[common.Hash]map[Listener]struct{} // contractAddress => logTopic => Listener
+		decoders      map[common.Address]AbigenContract
+		addSubscriber *utils.Mailbox
+		rmSubscriber  *utils.Mailbox
+		newLogs       *utils.Mailbox
+
 		utils.StartStopOnce
 		utils.DependentAwaiter
+		chStop chan struct{}
+		chDone chan struct{}
 	}
 
 	// The Listener responds to log events through HandleLog, and contains setup/tear-down
 	// callbacks in the On* functions.
 	Listener interface {
 		OnConnect()
 		OnDisconnect()
-		HandleLog(lb Broadcast, err error)
+		HandleLog(b Broadcast)
 		JobID() models.JobID
 		JobIDV2() int32
 		IsV2Job() bool
 	}
 
+	Config interface {
+		BlockBackfillDepth() uint64
+		TriggerFallbackDBPollInterval() time.Duration
+	}
+
+	ListenerOpts struct {
+		Contract AbigenContract
+		Logs     []generated.AbigenLog
+	}
+
 	AbigenContract interface {
 		Address() common.Address
-		ParseLog(log types.Log) (interface{}, error)
+		ParseLog(log types.Log) (generated.AbigenLog, error)
 	}
 
-	Config interface {
-		BlockBackfillDepth() uint64
-		TriggerFallbackDBPollInterval() time.Duration
+	registration struct {
+		listener Listener
+		opts     ListenerOpts
 	}
 )
 
 var _ Broadcaster = (*broadcaster)(nil)
 
 // NewBroadcaster creates a new instance of the broadcaster
 func NewBroadcaster(orm ORM, ethClient eth.Client, config Config) *broadcaster {
-	var (
-		dependentAwaiter = utils.NewDependentAwaiter()
-		relayer          = newRelayer(orm, config, dependentAwaiter)
-		subscriber       = newSubscriber(orm, ethClient, config, relayer, dependentAwaiter)
-	)
 	return &broadcaster{
-		subscriber:       subscriber,
-		relayer:          relayer,
-		DependentAwaiter: dependentAwaiter,
+		orm:              orm,
+		ethClient:        ethClient,
+		config:           config,
+		connected:        abool.New(),
+		registrations:    make(map[common.Address]map[common.Hash]map[Listener]struct{}),
+		decoders:         make(map[common.Address]AbigenContract),
+		addSubscriber:    utils.NewMailbox(0),
+		rmSubscriber:     utils.NewMailbox(0),
+		newLogs:          utils.NewMailbox(10000),
+		DependentAwaiter: utils.NewDependentAwaiter(),
+		chStop:           make(chan struct{}),
+		chDone:           make(chan struct{}),
```

Comment:
nit: but I think making this a wait group instead feels like a better (and more extensible) pattern. Add a `wg.Wait()` in `Stop()`, a `wg.Add(1)` in `Start()` and a `defer wg.Done()` in `awaitInitialSubscribers()`